recovery/elliptics_recovery/types/merge.py

```diff
@@ -733,114 +270,153 @@ def recover(self, keys):
         '''
         log.info("Server-send bucket: num keys: {0}".format(len(keys)))
 
-        def contain(key, ranges):
-            index = bisect(ranges, key)
-            return index % 2 == 1
+        def contain(address, backend_id, key):
+            addr, _, backend = self.routes.get_id_routes(key)[0]
+            return address == addr and backend_id == backend
 
-        responses = dict([(str(k), []) for k in keys]) # key -> [list of responses]
-        for addr, backend_id, backend_ranges in self.routes:
-            key_candidates = [k for k in keys if not contain(k, backend_ranges)]
-            if key_candidates:
-                self._server_send(key_candidates, addr, backend_id, responses)
+        responses = {str(k): [] for k in keys} # key -> [list of responses]
+        for addr, backend_id in self.backends:
+            key_candidates = [k for k in keys if not contain(addr, backend_id, k)]
+            for i in range(self.ctx.attempts):
+                if key_candidates:
```

Comment:
Default attempts value is 1. Usually attempts value is very small. So useless loop is not a big problem here.